internal/dataplane/sendconfig/inmemory.go

```diff
@@ -48,16 +54,33 @@ func (s UpdateStrategyInMemory) Update(
 
 	config, err := json.Marshal(targetState)
 	if err != nil {
-		return fmt.Errorf("constructing kong configuration: %w", err)
+		return fmt.Errorf("constructing kong configuration: %w", err), nil, nil
 	}
 
-	if _, err = s.configService.ReloadDeclarativeRawConfig(ctx, bytes.NewReader(config), true, false); err != nil {
-		return err
+	flattenErrors := shouldUseFlattenedErrors()
+	if errBody, err := s.configService.ReloadDeclarativeRawConfig(ctx, bytes.NewReader(config), true, flattenErrors); err != nil {
+		resourceErrors, parseErr := parseFlatEntityErrors(errBody, s.log)
+		return err, resourceErrors, parseErr
 	}
 
-	return nil
+	return nil, nil, nil
 }
 
 func (s UpdateStrategyInMemory) MetricsProtocol() metrics.Protocol {
 	return metrics.ProtocolDBLess
 }
+
+// shouldUseFlattenedErrors verifies whether we should pass flatten errors flag to ReloadDeclarativeRawConfig.
+// Kong's API library combines KVs in the request body (the config) and query string (check hash, flattened)
+// into a single set of parameters: https://github.com/Kong/go-kong/pull/271#issuecomment-1416212852
+// KIC therefore must _not_ request flattened errors on versions that do not support it, as otherwise Kong
+// will interpret the query string toggle as part of the config, and will reject it, as "flattened_errors" is
+// not a valid config key. KIC only sends this query parameter if Kong is 3.2 or higher.
+func shouldUseFlattenedErrors() bool {
+	return !versions.GetKongVersion().MajorMinorOnly().LTE(versions.FlattenedErrorCutoff)
+}
+
+type InMemoryClient interface {
+	BaseRootURL() string
+	ReloadDeclarativeRawConfig(ctx context.Context, config io.Reader, checkHash bool, flattenErrors bool) ([]byte, error)
+}
```

Comment:
To prevent copy pasta we can reuse the part of the interface that's already defined in https://github.com/Kong/kubernetes-ingress-controller/blob/35218dceeec9124319ea4cc49dfdc1ac2a468f39/internal/dataplane/sendconfig/inmemory.go#L16-L23 WDYT?

Then we could change this definition into 

```
type InMemoryClient interface{
	ConfigService
	BaseRootURL() string
}
```

WDYT?

Reply:
Makes sense, doing separately to just avoid more conflicts on the rest.